src/pipeline/comms/channel.rs

```diff
@@ -0,0 +1,141 @@
+//!
+//! # Examples
+//!
+//! ```
+//! # use raw_ipa::pipeline::comms::{Channel, Comms, Target};
+//! # use raw_ipa::proto;
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! // helper function that sets up 3 helpers hooked up to each other with channels
+//! let (c1, c2, c3, c_run) = raw_ipa::pipeline::util::intra_process_comms();
+//! tokio::spawn(c_run); // this initializes all of the runtime pieces for channels
+//!
+//! let message = String::from("hello");
+//! c1.send_to(Target::Next, proto::pipe::ExampleRequest { message: message.clone() }).await?;
+//! let recvd = c2.receive_from::<proto::pipe::ExampleRequest>().await?;
+//! assert_eq!(message, recvd.message);
+//! # Ok(())
+//! # }
+//! ```
+
+use crate::pipeline::buffer;
+use crate::pipeline::comms::{Comms, Target};
+use crate::pipeline::Result;
+use async_trait::async_trait;
+use log::{debug, error, info};
+use prost::Message;
+use serde::{Deserialize, Serialize};
+use std::io::Cursor;
+use std::time::Duration;
+use tokio::sync::{mpsc, oneshot};
+use tokio::time;
+use uuid::Uuid;
+
+pub struct Channel {
+    name: &'static str,
+    next_send_chan: mpsc::Sender<Vec<u8>>,
+    prev_send_chan: mpsc::Sender<Vec<u8>>,
+    buffer: mpsc::Sender<buffer::Command>,
+    shared_id: Uuid,
+}
+impl Channel {
+    #[must_use]
+    pub fn new(
+        name: &'static str,
+        next_send_chan: mpsc::Sender<Vec<u8>>,
+        prev_send_chan: mpsc::Sender<Vec<u8>>,
+        buffer: mpsc::Sender<buffer::Command>,
+        shared_id: Uuid,
+    ) -> Channel {
+        Channel {
+            name,
+            next_send_chan,
+            prev_send_chan,
+            buffer,
+            shared_id,
+        }
+    }
+}
+
+impl Drop for Channel {
+    fn drop(&mut self) {
+        info!("{} comms closing", self.name);
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct ChannelMessage {
+    shared_id: Uuid,
+    buf: Vec<u8>,
+}
+
+#[async_trait]
+impl Comms for Channel {
+    async fn send_to<M: Message>(&self, target: Target, data: M) -> Result<()> {
+        let mut buf = Vec::with_capacity(data.encoded_len());
+        // unwrap is safe because `buf` has `encoded_len` reserved
+        data.encode(&mut buf).unwrap();
+        let chan_message = ChannelMessage {
+            shared_id: self.shared_id(),
+            buf,
+        };
+        let res = serde_json::to_vec(&chan_message)?;
+        let chan = match target {
+            Target::Next => &self.next_send_chan,
+            Target::Prev => &self.prev_send_chan,
+        };
+        chan.send(res).await?;
+        Ok(())
+    }
+
+    async fn receive_from<M: Message + Default>(&self) -> Result<M> {
+        // basic poll for now; will use watchers in real implementation
+        loop {
+            let (tx, rx) = oneshot::channel();
+            self.buffer
+                .send(buffer::Command::Remove(self.shared_id(), tx))
```

Comment:
There might be a ticking bomb here - as long as `key` in the cache is `Copy` this model works fine. If `key` is composite, an owned value needs to be send down the channel every time cache is polled for data, i.e. a deep copy .

I wonder if we can spend some time working through some requirements for "buffer" layer - that would make it clear whether a given implementation is an incremental step towards our long term goal or not

Reply:
makes sense to me. tho I can definitely imagine that key will always be `Copy`. but we should definitely discuss